ts/nni_manager/common/gpu_scheduler/scheduler.ts

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  A simple GPU scheduler used by local and remote training services.
+ **/
+
+import { getLogger } from 'common/log';
+import { GpuSystemInfo, collectGpuInfo as origCollectGpuInfo } from './collect_info';
+
+const logger = getLogger('GpuScheduler');
+
+let collectGpuInfo = origCollectGpuInfo;  // for ut
+
+export interface ScheduleRestrictions {
+    onlyAcceptIndices?: number[];
+    rejectActiveGpus?: boolean;
+    rejectComputeActiveGpus?: boolean;
+}
+
+interface SchedulerGpuInfo {
+    index: number;
+    util: number;  // theoretical utilization calculated by NNI's trialGpuNumber
+    coreUtil: number;  // real GPU core utilization (0 ~ 1)
+    memUtil: number;  // real GPU memory utilization (0 ~ 1)
+    active: boolean;
+    computeActive: boolean;
+}
+
+interface SchedulerTrialInfo {
+    gpuIndex: number;
+    experimentId: string;
+    trialId: string;
+    util: number;
+}
+
+export class GpuScheduler {
+    private gpus: SchedulerGpuInfo[] = [];
+    private trials: SchedulerTrialInfo[] = [];
+
+    /**
+     *  Initialize the scheduler.
+     *
+     *  If there is no GPU found, throw error.
+     **/
+    public async init(): Promise<void> {
+        const info = await collectGpuInfo(true);
+        if (info === null) {
+            throw new Error('GpuScheduler: Failed to collect GPU info');
+        }
+        if (info.gpuNumber === 0) {
+            throw new Error('GpuScheduler: No GPU found');
+        }
+
+        for (let i = 0; i < info.gpuNumber; i++) {
+            this.gpus.push({
+                index: i,
+                util: 0,
+                coreUtil: 0,
+                memUtil: 0,
+                active: false,
+                computeActive: false,
+            });
+        }
+
+        this.updateGpus(info);
+    }
+
+    /**
+     *  Update GPUs' utilization info.
+     *
+     *  If `force` is not true, it may use cached result.
+     *
+     *  This is normally unnecessary because it will implicitly update before scheduling.
+     **/
+    public async update(force?: boolean): Promise<void> {
```

Comment:
Who will call this?

Reply:
GPU scheduler will be a shared daemon process. For example when a new experiment connects to the scheduler it will request a full update in it's start().
And I plan to show machine info on web ui. The user will be able to request an update.